Importer typehints and small importer refactor

semohr opened the pull request:
## Description

Hello y'all. 

One of the dev from the [beets-flask app](https://github.com/pSpitzner/beets-flask) here. We are extending the `ImportSession` in our application to get the interactive imports working in our application. Overall that was not a too pleasant experience as a ton of typehints are missing in that area of the beets code.  This PR wants to fix this ;)

### Changes

Moved importer state logic into a dataclass. Is clearer and more isolated now in my opinion, we are not using it but I found it quite confusing when looking at it.

Added a ton of typehints to the `importer.py` file. Predominantly the `ImportSession`, `ImportTask` (and derivatives) and pipeline stage decorators are now (hopefully) completely typed. Additionally I fixed some typhint issues in the `util.__init__.py` file, mainly related to the `PathLike` type. 

If you have some strong opinions about any of the changes feel free to revert or edit the PR.

Best,
Sebastian

P.S: I noticed that beets does not use Abstract Base Classes is there any reason for that or is it just of historic nature? I think some Base Classes could be adapted to use an ABC metaclass quite easily.


## To Do

- [x] Changelog. (Add an entry to `docs/changelog.rst` to the bottom of one of the lists near the top of the document.)
- [x] Tests finish successfully locally

semohr commented:
There is one failing tests that has, as far as I can see, nothing to do with any of the changes in the PR.  Some help would be highly appreciated here.

snejus commented:
I'm now looking at the test as that's something that I merged earlier, apologies!

snejus commented on test/plugins/test_lyrics.py:
```diff
@@ -22,7 +22,12 @@
-from beetsplug import lyrics
+
+try:
+    from beetsplug import lyrics
```
Lyrics integration tests only run if there have been adjustments in one of the lyrics source files.

Did you add this because these tests have been failing locally?

semohr replied:
Indeed, langdetect is not installed by poe. Atleast it wasn't for me. Therefore the tests fail.

I thought it is reasonable to just skip it, if the optional dependency is not installed.

I can revert the change if you want, dont mind either way.

snejus replied:
I see what you mean. There's a small issue here - we want the opposite in CI: if any dependencies are missing it should fail, forcing us to fix our CI setup and installing the missing dependency.

See `test_autobpm.py` for an example how this could be solved.

semohr commented:
I just merged the master into this branch, but the test still seems to be broken. Tried to fix it but tbh I have no idea what the issue is here, as I don't have much insights into how the lyrics plugin works. It is also difficult to debug as the issue does not occur locally for me. From my side feel free to merge this!

I might make another PR in the near future for some changes related to the pipeline and internals.

snejus commented:
Apologies, I haven't yet addressed the lyrics issue... 🤦🏼

snejus commented:
https://github.com/beetbox/beets/pull/5618

semohr commented:
Don't stress it! Was just checking it again and wasn't able to fix it myself.

snejus commented:
It's been merged - feel free to rebase / merge now! Will have a look at the types you added now, coincidentally I've been doing the same in my fork, so I'm intrigued! 😉

semohr commented:
Awesome! If you want feel free to check my [async-import](https://github.com/semohr/beets/tree/async-import) branch too. I added some more type hints there too, will maybe make a PR once this is a bit more polished and it is fully backwards compatible.

snejus commented on beets/importer.py:
```diff
@@ -1268,11 +1336,11 @@ def tasks(self):
         # the extracted directory).
         if self.is_archive:
-            yield archive_task
+            yield archive_task  # type: ignore
```
If we _must_ have this in place, it needs a specific error type being ignored. Is it actually needed?

semohr replied:
You are right not needed anymore

snejus reviewed (changes requested):
Added a couple of comments, note I will have another look once these are addressed, thanks!

semohr commented on beets/importer.py:
```diff
@@ -1,22 +1,9 @@
-# a copy of this software and associated documentation files (the
-# "Software"), to deal in the Software without restriction, including
-# without limitation the rights to use, copy, modify, merge, publish,
-# distribute, sublicense, and/or sell copies of the Software, and to
```
It is not necessary since the root of the project contains the license. Adds quite some bloat imo, also it is in the place where someone would expect a module level docstring. 
I can add it again if you feel strongly about it, but it I feel it is completely unnecessary.

snejus replied:
If we are to remove this, we probably want to do this in a separate PR across the entire codebase. This is simply out of scope for the work that you have in this PR

semohr commented on beets/importer.py:
```diff
@@ -25,15 +12,17 @@
+from dataclasses import dataclass
 from enum import Enum
 from tempfile import mkdtemp
+from typing import Callable, Dict, Iterable, Optional, Sequence, Union, cast
```
The typing Union is not getting deprecated. You would not need to drop it once you migrate to python 3.10 but sure thing will change it.

semohr replied:
FYI: The tests for python 3.9 fail now as pipe unions are not supported in every context, welp

semohr commented on beets/util/__init__.py:
```diff
@@ -198,8 +198,8 @@ def ancestry(path: AnyStr) -> list[AnyStr]:
-    path: AnyStr,
-    ignore: Sequence[bytes] = (),
+    path: PathLike,
+    ignore: Sequence[PathLike] = (),
```
It is converted to a bytes_path before passing to the fnmatch function tho. See line 230. I don't see a problem here even tho mypy raises it. Can you elaborate where the problem occurs here?

snejus replied:
No issues regarding `path` - I see it's converted properly and `mypy` is happy about it. Instead, the issue has to do with the `ignore` parameter: 

```python
...
        for pat in ignore:
            if fnmatch.fnmatch(base, pat):  # mypy: Value of type variable "AnyStr" of "fnmatch" cannot be "str | bytes | Path"
...
```

And my wider question is whether `sorted_walk` ever receives `pathlib.Path` type?

semohr replied:
I figured it out. There is no problem here! Mypy does not allow for variable shadowing if one renames the `ignore` variable as it is typed as `[PathLike]` mypy has no issues anymore.

snejus replied:
I'm not yet convinced this function needs to support `pathlib.Path`. The only place where it's used in this PR is here:

```python
    ignore = cast(list[str], config["ignore"].as_str_seq())
    ignore_hidden = cast(bool, config["ignore_hidden"].get(bool))

    for root, dirs, files in sorted_walk(
        path, ignore=ignore, ignore_hidden=ignore_hidden, logger=log
    ):
```

As far as I can see, `ignore` is a list of strings?

semohr replied:
I don't think supporting it adds too much runtime and it improves readability and re-usability. Your call tho :)
```
    _ignore = [  # rename prevents mypy variable shadowing issue
        bytestring_path(i) for i in ignore
    ]
```

snejus replied:
For me, a 'tighter' type is always easier to handle and understand (e.g. just `bytes` vs `bytes | str | pathlib.Path`).

> _ignore = [  # rename prevents mypy variable shadowing issue
>         bytestring_path(i) for i in ignore
>     ]

This line adds complexity to the code base, so if I looked at this function tomorrow and wanted to simplify it, I'd find that it doesn't need to support `pathlib.Path` and I'd end up removing it, since no code ever calls it with `pathlib.Path`.

semohr replied:
It does but I didn't add it, it was already present. Not sure if removing it breaks something for third party code, as it is not marked as private.

snejus replied:
Could you just rename it to `ignore_bytes` for clarity?

semohr commented on beets/importer.py:
```diff
@@ -1542,7 +1616,7 @@ def emitter(task):
             group_albums(session),
             lookup_candidates(session),
-            user_query(session),
+            user_query(session),  # type: ignore # Recursion in decorators
```
Indeed this was a fix for a early problem and is not required anymore.

semohr commented on beets/importer.py:
```diff
@@ -25,15 +12,17 @@
 from beets import autotag, config, dbcore, library, logging, plugins, util
 from beets.util import (
     MoveOperation,
+    PathLike,
```
Replaced it with a `PathBytes` type which mirrors bytes for now. May allow to migrate to a PathLike easier in the future.

snejus commented on beets/importer.py:
```diff
@@ -706,7 +753,9 @@ def find_duplicates(self, lib):
         # use a temporary Album object to generate any computed fields.
         tmp_album = library.Album(lib, **info)
-        keys = config["import"]["duplicate_keys"]["album"].as_str_seq()
+        keys = cast(
```
Here, I think, you can just use `keys: list[str] = ...`

semohr replied:
Sadly this raises a mypy error. As far as I can see you always need to cast the return values from the confuse helpers.
```
Type "Unknown | object | Path | AttrDict | None" is not assignable to declared type "list[str]"
  Type "Unknown | object | Path | AttrDict | None" is not assignable to type "list[str]"
    "object" is not assignable to "list[str]"
```

snejus replied:
Interesting, I'm not seeing the same issue. As far as I can see, `confuse` doesn't have any type annotations:

```python
    def as_str_seq(self, split=True):
        """Get the value as a sequence of strings. Equivalent to
        `get(StrSeq(split=split))`.
        """
        return self.get(templates.StrSeq(split=split))
```

Could you try to assign it this way and push the code, to see whether the type check fails in CI?

```python
        keys: list[str] = config["import"]["duplicate_keys"][
            "item"
        ].as_str_seq()
```

semohr replied:
I had pylance enabled in my setup. Still, the ground issue is suppressed by mypy since it assumes all confuse types are Any since it is set to ignore it. Pylance (typechecker) on the other side tries to infer the type from external packages which will raise an error as the return types in confuse are not set and can't be inferred correctly. Adding proper type hints to the confuse package will increase ux, pylance is the default type checker for many people.

For example take the as_str_seq function:
```python
def as_str_seq(self, split=True) : # -> Any | object | None | Path | AttrDict:
```
Should be 
```python
def as_str_seq(self, split=True)  -> Any :
    pass
# or even better
def as_str_seq(self, split=True)  -> Sequence[str] :
    pass
```

snejus replied:
In my opinion, in this case `mypy` behaviour is more tolerable: it provides the flexibility to assign the type on our end (`keys: list[str] = ...`) without having to depend on `cast`. This means that once types are added in confuse, and if they do not align with what we've assigned, the type checker will fail and we will need to address the discrepancy.

On the other hand, such discrepancy will be silently ignored if the type was `cast`.

snejus replied:
The project uses mypy so developers are expected to use it instead of pylance. Though, I don't think our docs mention such expectation or mypy at all - do you reckon this would be helpful for future developers?

semohr replied:
I mean mypy is just skipping the check of external packages (as it is configured to do) that is okay for automated pipelines as we assume other packages are implemented properly, but for the developer it is quite useful to have some more interspersion. Thinking about possible type mismatches and debugging issues. I agree confuse should handle it properly, for now I will have to live with the local errors I guess.  

Normally pylance and mypy are not exclusive they work quite well together.  As far as I know, pylance comes with vscode in the default python environment , therefore many people just use it without thinking about it too much or even noticing the use it ;)

I think adding a small section on typing in the contribution guide would be useful especially if you want to enforce some typing rules for new code. Also I just noticed the contribution guide is a bit outdated.

snejus replied:
Well, you're welcome to switch to neovim and use mypy + jedi - this should make sure you don't have errors locally 😆 

I will update the guidelines accordingly. Regarding the comment style, do you mean the docstring style? Best comment style is no comments at all, where code is completely self-explanatory 😅

semohr replied:
I currently dont have the passion to learn all vim motions, but it is on my todo list :P

I think no comments is a bit idealistic, there will always be external developers who might need some more aid understanding the source. I like to sets a docstring format and add a ruff rule for at least requiring documenting public facing functions. This allows to reference them in sphinx quite easily and show their docstring, arguments, returns and types. This also works for modules and basically any object because of pythons introspection.

For instance some references could improve the "writing plugins" guide significantly.

[Ref](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html)

snejus replied:
From my experience dealing with outdated and misleading docstrings (that include input and output parameters), I found that appropriately named parameters with accurate types work best. Then the docstring can be used to summarise the purpose of the function in the first line, and expand in detail in another paragraph if needed.

snejus commented on beets/importer.py:
```diff
@@ -973,14 +1030,14 @@ def prune(self, filename):
                 os.path.dirname(filename),
                 self.toppath,
-                clutter=config["clutter"].as_str_seq(),
+                clutter=cast(list[str], config["clutter"].as_str_seq()),
```
Try to assign a variable instead of using `cast`

```python
clutter: list[str] = config["clutter"].as_str_seq()
```

snejus commented on beets/importer.py:
```diff
@@ -1022,7 +1083,9 @@ def find_duplicates(self, lib):
         # temporary `Item` object to generate any computed fields.
         tmp_item = library.Item(lib, **info)
-        keys = config["import"]["duplicate_keys"]["item"].as_str_seq()
+        keys = cast(
```
Same as above

snejus reviewed (changes requested):
Noticed that `typing.cast` is being used - unfortunately it tends to hide typing issues. Added some suggestions how to replace it

snejus commented on beets/importer.py:
```diff
@@ -25,9 +26,10 @@
+from dataclasses import dataclass
 from enum import Enum
 from tempfile import mkdtemp
+from typing import Callable, Iterable, Optional, Sequence, Union, cast
```
I'd expect the `Optional` import to go away, see 

![image](https://github.com/user-attachments/assets/98190248-d130-4394-b533-8da3df86dfbc)

semohr replied:
This is a bit of a style choice I guess. If we don't want to allow this globally there is an experimental ruff rule for it see https://docs.astral.sh/ruff/rules/non-pep604-annotation-optional/.

snejus replied:
To be fair that's the rule I used when I replaced all instances of `Optional` by the pipe-based annotations in #5508 but I forgot to uncomment it in `pyproject.toml`: 

```toml
[tool.ruff.lint]
select = [
    # "ARG", # flake8-unused-arguments
    # "C4", # flake8-comprehensions
    "E", # pycodestyle
    "F", # pyflakes
    # "B", # flake8-bugbear
    "I", # isort
    "N", # pep8-naming
    "PT", # flake8-pytest-style
    # "RUF", # ruff
    # "UP", # pyupgrade
    "TCH", # flake8-type-checking
    "W", # pycodestyle
]
```

semohr replied:
Just out of interest why do you have a dislike for Optional? It is not getting deprecated as far as I can see.

I like it, aa it is a bit more humanly readable for nested types.

snejus replied:
Interesting, readability is why I prefer pipe-based annotations: for example, they are less verbose thus less liable to be split across multiple lines by black. A couple of reasons I can think of:

1. As mentioned, I agree with [pyupgrade / ruff](https://docs.astral.sh/ruff/rules/non-pep604-annotation-optional/)
   > This syntax is more concise and readable than the previous typing.Optional syntax
2. It's in line with union annotations in other languages, for example lua and typescript
3. Doesn't require an additional import (except for `from __future__ import annotations` with Python < 3.10)
4. Less verbose

snejus replied:
On the same note `ruff` can help you address this:

```sh
ruff check --select=UP045,F401 --preview --unsafe-fixes --fix
```

semohr replied:
Strangely I mostly don't agree with that. Here is my take on it, if you care about it :D

1. I always read from left to right and if a value may be None is one of the most important things to know. So The Optional keyword is the first thing I want to see. The problem with pipe based syntax is that it does not have to be sorted. thus the effort to determine if a value may be None is higher for me,
2. Coming from rust and haskell I am used to the `Option` type, maybe I'm biased here. 
3. Simplicity is a good thing, but from a verbosity standpoint the Optional keyword is only three character longer than the pipe syntax with None. ` | None` vs `Optional[]`. Doesn't make much of a different for me and one can always mirror the type with e.g. an `O[]` or ` | N` which is four / three characters shorter.

Nonetheless, I replaced the Optional type with the pipe tho. FYI the currently pinned ruff version doesn't support `UP045` yet.

semohr commented on beets/importer.py:
```diff
@@ -73,149 +60,174 @@
-    return toppath in state
-
+    def __init__(self, readonly=False, path: Optional[PathBytes] = None):
+        self.path = path or cast(PathBytes, config["statefile"].as_filename())
```
Same, I do not like it too much. Sadly there are some issues with that are just easier to solve with it. Mypy sometimes forces questionable patterns otherwise.

The issue here is that the typehints from confuse are not really that useful and mypy will not infer as_filename as str (which I assume it is?). A cast is preferred here if we don't want to introduce another variable or union. 
```python
os.fsencode(cast(str, config["statefile"].as_filename()))
```

snejus commented on beets/util/pipeline.py:
```diff
@@ -149,7 +152,17 @@ def multiple(messages):
 
 
-def stage(func):
+A = TypeVarTuple("A")
```
Would it be possible to document what do these variables represent? I am a bit confused what do they stand for and how they relate to each other. mypy also complains about incompatible type on line 183

semohr replied:
How about renaming them, should make it more clear. TypeVarTuple and TypeVar are generics. [TypeVarTuple](https://peps.python.org/pep-0646/) basically allows for generics arguments with typed elements. Whereas TypeVar is just a basic generic. We combine the Args generic with the Task generic i.e. (*args, task) for properly typing the callable function argument of the generator.

```
# Arguments of the function (omitting the task)
Args = TypeVarTuple("Args")
# Task as an additional argument to the function
Task = TypeVar("Task")
# Return type of the function (should normally be task but sadly
# we cant enforce this with the current stage functions without
# a refactor)
R = TypeVar("R")
```

The mypy error is again a shadowing problem which is easy to resolve no real issue as far as I can see.

snejus replied:
That's nice, thanks. Feel free to follow the standard notation with variables `A` for args and `T` for task - I just needed a pair of comments there really

snejus commented:
See the last two bits in unresolved conversations - otherwise we'll be good to go!

snejus commented on beets/importer.py:
```diff
@@ -1731,10 +1834,10 @@ def group(item):
         for _, items in itertools.groupby(sorted_items, group):
-            items = list(items)
-            task = ImportTask(task.toppath, [i.path for i in items], items)
+            l_items = list(items)
```
This is an interesting one. It seems like originally it was broken: the `items` iterator is emptied by the time it's sent into `ImportTask`. Is that correct?

semohr replied:
Indeed this could have happen if it truely was an iterator (which it is mostly not), this was an easy catch with the additions to the type hints.

snejus replied:
As far as I know, `groupby` _always_ returns at iterator as the second returned item. This kind of seems to me that `SingletonimportTask` never ends up using the provided `items`.

snejus reviewed (changes requested):
I realised I haven't had a look at the changes that this PR is actually about 🤦🏼 I did not notice that GitHub hid the diff initially.

While most of the changes add types, I noticed that in a couple of places there are some updates in functionality. Are we sure that those bits are covered by tests?

I noticed there's a pattern of using `assert` against some 'no-go' cases. I don't think we want to have them since in practice they never happen. On the other hand, I see how those bits are ambiguous: this is a symptom of a badly designed `ImportTask` which should have been split into `AlbumImportTask` and `SingletonImportTask`. Ideally we should document these shortcomings in docstrings or TODOs in the comments instead.

snejus commented on beets/importer.py:
```diff
@@ -366,26 +400,28 @@ def run(self):
+
     @property
     def history_dirs(self):
-        if not hasattr(self, "_history_dirs"):
-            self._history_dirs = history_get()
+        if self._history_dirs is None:
+            self._history_dirs = ImportState().taghistory
         return self._history_dirs
```
Feel free to ignore this - I see this is outside of the scope of your change.

semohr replied:
Added a fixme and typehints.

semohr commented on beets/importer.py:
```diff
@@ -73,149 +74,174 @@
 
+    def __init__(self, readonly=False, path: PathBytes | None = None):
+        self.path = path or os.fsencode(config["statefile"].as_filename())
+        self.tagprogress = {}
```
You are right we could use the fields feature here. I normally tend to disable the automatic init functions for dataclasses, i.e. `@dataclass(init=False)` as this makes it a bit more clear for me what is happening. Is mostly a style thing, I can change it to whatever you feel fits.

snejus replied:
That's the main feature that I use them for! 😅 I usually use them to define the properties that the object is expected to have, and let it do its job regarding object initialisation. If something extra needs to be done during init, I don't override `__init__` and instead use a factory method, for example:

```python
class ExecutionContext(NamedTuple):
    """Context details for the current Lambda execution."""

    functionality: str
    integration: str | None
    region: str

    @classmethod
    def from_arn(cls, arn: str) -> Self:
        """Parse relevant parts from the lambda ARN.

        Lambdas names follow two patterns:
        - client-specific: <function>-<integration>-<env>
        - client-agnostic: <function>-<env>
        """
        parts = arn.split(":")[-1].split("-")[:-1]
        return cls(
            functionality=parts[0],
            integration=parts[1] if len(parts) > 1 else None,
            region="uk" if ":eu-west-" in arn else "us",
        )
```

I found this pattern makes it easy to test the custom construction: 

```python
assert ExecutionContext.from_arn("something") == ExecutionContext(**expected_data)
```

semohr replied:
I think for me this stems from experiences with the overuse or missuse of the __post_init__ dunder.  For more complex validation logic or default values which should always occur regardless of alternate constructors, I normally don't want to write another function to include as a default factory. 

```python
@dataclass
class Event:
    start_time: datetime
    end_time: datetime = field(default_factory=lambda: datetime.now() + timedelta(hours=1))

    def __post_init__(self):
        if self.end_time <= self.start_time:
            raise ValueError("end_time must be after start_time")
 ```
 vs
 ```python
@dataclass(init=False)
class Event:
    start_time: datetime
    end_time: datetime

    def __init__(self, start_time, end_time):
        self.start_time = start_time
        self.end_time = end_time or start_time + timedelta(hours=1)
        if self.end_time <= self.start_time:
            raise ValueError("end_time must be after start_time")
 ```
The second example is more clear to me and should also be a bit more performant as there are less function calls.
 
I mainly use dataclasses as they are serializable with `asdict` and automatically implement hashing and equality checks.

To get back to the initial issue, I don't care either way and it is really not that important to me. I can change it to whatever fits the style of the existing code :grinning:

snejus replied:
Feel free to keep it as is since it's more of a nit! You've already addressed just about a hundred of comments here 😅

semohr commented on beets/importer.py:
```diff
@@ -73,149 +74,174 @@
+            A logging handler to use for the session's logger. If None, a
+            NullHandler will be used.
+        paths : os.PathLike or None
+            The paths to be imported. If None, no paths are specified.
```
Good point, this should go. Im actually not sure why the paths can be None, I was looking for some examples where this might happen but didn't find any. Decided to keep it as there might be an edge case I missed. It was like that in the pre-pr state

semohr commented:
I did not change any of the functionality, I just moved the state read and write functions into a dataframe and added a context manager. I do think the existing tests should cover this functionality. The test coverage at least seems to still catch most things, the only function not tested is the progress_reset one, which wasn't tested beforehand afaiks.

semohr commented on beets/importer.py:
```diff
@@ -1267,12 +1359,12 @@ def tasks(self):
         # for archive imports, send the archive task instead (to remove
         # the extracted directory).
-        if self.is_archive:
+        if archive_task:
```
We do the check this early and the `archive_task!=None` if it is an archiveimporttask. If we don't do it this way mypy will complain and not infer the yield type correctly.

snejus replied:
Consider simplifying this to

```python
        yield archive_task or self.sentinel()
```

semohr commented on beets/importer.py:
```diff
@@ -1425,12 +1518,16 @@ def _extend_pipeline(tasks, *stages):
     import, yields single-item tasks instead.
     """
     skipped = 0
+    if session.paths is None:
```
Would be preferable in my opinion too. I didn't want to introduce too many logic changes but tbh we went down that road anyways so no harm adding this :)

semohr commented on beets/importer.py:
```diff
@@ -366,26 +400,28 @@ def run(self):
         """
         if self.is_resuming(toppath) and all(
-            [progress_element(toppath, p) for p in paths]
+            [ImportState().progress_has_element(toppath, p) for p in paths]
```
At them moment yes that is how it works yes. We can attach one of the objects to the session if we want to, we do something similarly in our overwrites to the session in the beets-flask app. This would need quite some changes tho. Maybe a fixme for now?

snejus replied:
Yup, definitely out of scope for this PR! This will also need some care, I remember refactoring this where I initialized the state once for a session - next thing I knew my state file was completely emptied out 😓

semohr commented on beets/importer.py:
```diff
@@ -403,30 +439,30 @@ def mark_merged(self, paths):
                 log.warning(
                     "Resuming interrupted import of {0}",
-                    util.displayable_path(toppath),
+                    util.displayable_path(normpath(toppath)),
```
Yes that would work, but might be reverted if we move to allow Path objects here. In my opinion only the db and file system functions really need to operate on bytes. Or do I miss something? Feels a bit out of scope for me.

snejus replied:
> only the db and file system functions really need to operate on bytes

I think these two cover a large majority of how paths end up being used. More importantly though, I think we need consistency, to make our life easier when we eventually migrate to `pathlib.Path`.

I found this to be in scope of this PR because these `normpath` calls did not exist before.

semohr commented on beets/importer.py:
```diff
@@ -492,48 +537,63 @@ class ImportTask(BaseImportTask):
         ):
-            self.choice_flag = choice
+            # Cast needed as mypy can't infer the type
+            self.choice_flag = cast(action, choice)
```
Fully agree!

semohr commented on beets/importer.py:
```diff
@@ -568,26 +628,32 @@ def imported_items(self):
 
     def apply_metadata(self):
         """Copy metadata from match info to the items."""
+        assert isinstance(
+            self.match, autotag.AlbumMatch
+        ), "apply_metadata() only works for albums"
+
```
This will introduce some mypy issues tho. I guess we are fine with that?

snejus replied:
I think mypy may be happy with this. It would be good to also remove the assert from `SingletonImportTask.apply_metadata` too

semohr commented on beets/importer.py:
```diff
@@ -764,19 +832,25 @@ def align_album_level_fields(self):
 
         If `write` is `True` metadata is written to the files.
         """
+        assert session is not None, "session must be provided"
```
I added a todo: # TODO: Intorduce a MoveOperation.NONE or SKIP

snejus replied:
Amazing, thanks!

semohr commented on beets/importer.py:
```diff
@@ -806,15 +880,17 @@ def manipulate_files(self, operation=None, write=False, session=None):
             self.album = lib.add_album(self.imported_items())
-            if self.choice_flag == action.APPLY:
+            if self.choice_flag == action.APPLY and isinstance(
+                self.match, autotag.AlbumMatch
```
Should we keep it for now? Otherwise I guess this will yield some mypy errors.

snejus replied:
I see what you mean. Let's keep this in place then

semohr commented on beets/importer.py:
```diff
@@ -995,13 +1071,17 @@ def chosen_info(self):
-        elif self.choice_flag is action.APPLY:
+        elif self.choice_flag is action.APPLY and self.match:
             return self.match.info.copy()
+        assert False
```
I think the issue here is that match is None if the action is not APPLY. There is no way to join these two variables without a redesign of the data structure. Therefore I added the assert to allow mypy to infer the proper return type. I can raise an error otherwise but just removing the elif wont work if we want to keep proper typehints.

snejus replied:
In this case I think I'd rather just slap another `type: ignore` with an explanation what's going on and that it needs fixing:

```python
    def chosen_info(self) -> dict[str, Any]:
        """Return a dictionary of metadata about the current choice.
        May only be called when the choice flag is ASIS or RETAG
        (in which case the data comes from the files' current metadata)
        or APPLY (in which case the data comes from the choice).
        """
        if self.choice_flag in (action.ASIS, action.RETAG):
            likelies, consensus = autotag.current_metadata(self.items)
            return likelies

        # Otherwise the action is APPLY, which means the match is set
        # TODO: redesign to make this explicit and type safe
        return self.match.info.copy()  # type: ignore[union-attr]
```

Note I added the output type. By default `mypy` doesn't check untyped functions.

semohr commented on beets/importer.py:
```diff
@@ -1092,23 +1174,24 @@ def save_history(self):
 
     def save_progress(self):
-        if self.paths is None:
+        if self.paths is None and self.toppath:
```
The progress reset function wont work if no toppath is given. This check just ensures that we dont run into this issue as in theory it can happen. Maybe we want to raise an error here if paths and toppath is none.

snejus replied:
It seems like progress reset checks for existence before attempting to remove:

```python
            if toppath in state.tagprogress:
                del state.tagprogress[toppath]
```

Seems to me should be happy handling `None` values? It's been working fine, and the implementation of progress reset seems to have stayed more or less the same?

snejus replied:
Additional comment: `self.paths` is always a list I think, so you may want to use `if not self.paths`

snejus reviewed (approved):
Thanks for your patience - we're good to go. Just rebase the latest master and we can merge it.
